crates/macros/src/lib.rs

```diff
@@ -424,3 +443,82 @@ pub fn test_scripts(input: proc_macro::TokenStream) -> TokenStream {
+///
+/// **Note** When using `random!` you should use `all_predicates` in the output to ensure
+/// that the test does not fail due to events happening out of order!
+///
+/// Usage:
+///
+///   `run_test[inputs, script1, script2, ...]`
+///
+/// The generated test will:
+///   - take the first entry of `inputs`, which should be a `Vec<HotShotEvent<TYPES>>`,
+///   - if the input is random, it'll generate a `seed` and shuffle the inputs.
+///   - if the input is serial, it will execute in order.
+///   - print the seed being used.
+///   - feed this to each task in order, validating any output received before moving on to the next one,
+///   - repeat the previous steps, with the aggregated outputs received from all tasks used as the new input set,
+///   - repeat until no more output has been generated by any task, and finally
+///   - proceed to the next entry of inputs.
+///   - print the seed being used (again) to make finding it a bit easier.
+///
+/// # Panics
+///
+/// The macro panics if the input stream cannot be parsed.
+/// The test will panic if the any of the scripts has a different number of stages from the input.
+#[proc_macro]
+pub fn run_test(input: TokenStream) -> TokenStream {
+    // Parse the input as an iter of Expr
+    let inputs: Vec<_> = syn::parse::Parser::parse2(
+        syn::punctuated::Punctuated::<syn::Expr, syn::Token![,]>::parse_terminated,
+        input.into(),
+    )
+    .unwrap()
+    .into_iter()
+    .collect();
+
+    // Separate the first input (which should be the InputOrder enum)
+    let test_inputs = &inputs[0];
+    let scripts = &inputs[1..];
+
+    // Generate code for shuffling and flattening inputs
+    let expanded = quote! {
+        {
+            use rand::{
+                SeedableRng, rngs::StdRng,
+                seq::SliceRandom
+            };
+            use hotshot_task_impls::events::HotShotEvent;
+            use hotshot_task::task::TaskState;
+            use hotshot_types::traits::node_implementation::NodeType;
+            use hotshot_testing::script::InputOrder;
+
+            async {
+                let seed: u64 = rand::random();
+                tracing::info!("Running test with seed {seed}");
+                let mut rng = StdRng::seed_from_u64(seed);
+                let mut shuffled_inputs = Vec::new();
+
+                for (stage_number, input_order) in #test_inputs.into_iter().enumerate() {
+                    match input_order {
+                        InputOrder::Random(mut events) => {
+                            events.shuffle(&mut rng);
+                            shuffled_inputs.push(events);
+                        },
+                        InputOrder::Serial(events) => {
+                            shuffled_inputs.push(events);
+                        }
+                    }
+                }
+
+                test_scripts![shuffled_inputs, #(#scripts),*].await;
+                tracing::info!("Suite used seed {seed}");
```

Comment:
It occurred to me -- this might not appear in the same conditions that the panic in `test_scripts` does. 

could we e.g. https://doc.rust-lang.org/std/panic/fn.catch_unwind.html the `test_scripts!` panic and then panic ourselves with the seed? this would avoid situations where the test fails but we don't see the seed because the log level was not high enough etc

Reply:
Spoke offline: TL;DR, this is hard.